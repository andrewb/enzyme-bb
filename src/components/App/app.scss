.app {
  padding: 2em;

  &__cart {
    overflow: hidden;

    .toggle {
      float: right;
      text-align: right;

      &__content {
        margin-top: 0.5em;
        padding: 2em;
        background: #fff;
        text-align: left;
      }      
    }
  }
}
