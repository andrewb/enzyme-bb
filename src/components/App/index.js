import PropTypes from 'prop-types';
import React, { Component } from 'react';
import Inventory from '../Inventory';
import Toggle from '../Toggle';
import './app.scss';

class App extends Component {
  constructor(props) {
    super(props);
    this.state = {
      products: props.products,
      cart: 0
    };
    this.onAddToCart = this.onAddToCart.bind(this);
  }
  onAddToCart(sku) {
    // Add the product to the cart
    // This simply increments the cart count and decrements the quantity of the
    // chosen product.
    // NOTE, IRL use something like redux...
    this.setState({
      cart: this.state.cart + 1,
      products: this.state.products.map((p) => {
        if (p.sku === sku) {
          // Decrement product quantity
          return Object.assign({}, p, { quantity: p.quantity > 0 ? p.quantity - 1 : 0 });
        }
        return Object.assign({}, p);
      })
    });
  }
  render() {
    const props = {
      products: this.state.products.map(p => (
        Object.assign({}, p, { onAddToCart: this.onAddToCart })
      ))
    };
    return (
      <div className="app">
        <div className="app__cart">
          <Toggle label={`Cart (${this.state.cart})`}>
            <em>The cart would go here...</em>
          </Toggle>
        </div>
        <Inventory {...props} />
      </div>
    );
  }
}

App.defaultProps = {
  products: []
};

App.propTypes = {
  products: PropTypes.arrayOf(PropTypes.shape({
    sku: PropTypes.string,
    name: PropTypes.string,
    quantity: PropTypes.number,
    price: PropTypes.number,
    image: PropTypes.object
  }))
};

export default App;
