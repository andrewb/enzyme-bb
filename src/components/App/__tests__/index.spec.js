import React from 'react';
import renderer from 'react-test-renderer';
import App from '../index';

describe('<App />', () => {
  // NOTE: this is an example of a snapshot test.
  // See https://facebook.github.io/jest/docs/snapshot-testing.html
  it('renders correctly', () => {
    const tree = renderer.create(
      <App
        products={[
          {
            sku: 'xyz',
            name: 'Product Name',
            quantity: 10,
            price: 100,
            image: {
              src: 'https://foo.bar/baz.jpg'
            }
          }
        ]}
      />).toJSON();
    expect(tree).toMatchSnapshot();
  });
  // Compare this to using Enzyme to test that certain components are rendered.
  /*
  it('renders an <Inventory />', () => {
    const wrapper = shallow(<App />);
    expect(wrapper.find(Inventory).exists()).toBe(true);
  });
  it('renders a cart <Toggle />', () => {
    const wrapper = shallow(<App />);
    expect(wrapper.find(Toggle).exists()).toBe(true);
  });
  */
  // I generally prefer the latter approach, because it works better with BDD
  // and IME is less fragile. However, both are useful â€“ it all depends on the
  // situation.
});
