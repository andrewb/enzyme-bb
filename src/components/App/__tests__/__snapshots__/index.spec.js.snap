// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<App /> renders correctly 1`] = `
<div
  className="app"
>
  <div
    className="app__cart"
  >
    <div
      className="toggle"
    >
      <div
        className="toggle__button toggle__button--is-closed"
      >
        <button
          className="button"
          onClick={[Function]}
          type="button"
        >
          <span
            className="toggle__label"
          >
            Show Cart (0)
          </span>
        </button>
      </div>
    </div>
  </div>
  <div
    className="inventory"
  >
    <header
      className="header"
    >
      <h1>
        Our Products
      </h1>
    </header>
    <ul
      className="inventory__list"
    >
      <li
        className="inventory__list-item"
      >
        <div
          className="product"
        >
          <div
            className="product__image"
          >
            <img
              alt="Product - Product Name"
              src="https://foo.bar/baz.jpg"
            />
          </div>
          <div
            className="product__details"
          >
            <h1
              className="product__name"
            >
              Product Name
            </h1>
            <div
              className="product__sku"
            >
              xyz
            </div>
            <div
              className="product__price"
            >
              $
              100.00
            </div>
            <div>
              <div
                className="product__availability"
              >
                <span
                  className="product__quantity"
                >
                  10
                </span>
                 in Stock
              </div>
              <div
                className="product__cart-button"
              >
                <button
                  className="button"
                  onClick={[Function]}
                  type="button"
                >
                  <span>
                    Add to Cart
                  </span>
                </button>
              </div>
            </div>
          </div>
        </div>
      </li>
    </ul>
  </div>
</div>
`;
